lib/hl7/datetime_components.rb

```diff
@@ -0,0 +1,64 @@
+require 'active_support/core_ext/time'
+
+module HL7
+  class DatetimeComponents
+    attr_reader :year, :month, :day, :hour, :minute, :second, :fraction, :offset_seconds
+
+    def initialize(year, month = nil, day = nil, hour = nil, minute = nil, second = nil, fraction = nil, offset_seconds: nil)
```

Comment:
This doesn't really enforce the intended invariant (which is that if a component is non-nil then all of the more significant components are also non-nil).

Do you think it's worth enforcing that explicitly?

Reply:
> Do you think it's worth enforcing that explicitly?

yes

Reply:
The idea here is that, if you want to know if a given datetime only has day-level precision, then you can do:
```ruby
components.hour.nil?
```
Well, that doesn't guarantee day-level precision; it just means we _don't_ have at least hour-level precision. We could define convenience methods for useful tests. Right now we don't do anything with precision information, but it could be useful in principle. If a source only gives us day-level precision for cancellation timestamps -- and one of them does exactly that -- we might want to store that somehow. 

Precision of time data in general (not just in HL7 datetime strings) is an issue in a number of places, like ER discharges, for example. For what that's worth...